### ISSUE TITLE
optimize values list execution plan by moving the evaluation part to execution phase

### ISSUE THREAD

**Issue by jimexist** [apache/datafusion#1169]
I wonder what you think about moving the creation of the actual `RecordBatch` from `ValuesExec::try_new` to `execute` -- the rationale would be to make `PhysicalPlan` creation faster and push the actual work into `execute` where if can potentially be run concurrently with other parts

Given the size of data in a `VALUES` statement, this is not likely to be any real difference so I am fine with leaving the creation in the same place too -- I just wanted to mention it.

_Originally posted by @alamb in https://github.com/apache/arrow-datafusion/pull/1165#r734957534_

**Comment by ygf11** [comment]
It seems statistics computation in ValuesExec still need the evaluation result. 
It is ok to ignore this by creating default statistics? or we need iterate data again to compute statistics?

**Comment by alamb** [comment]
I think don't think adding an RwLock sounds like a good idea. Perhaps we could compute the statistics when the `ValuesExec` is created (rather than computing on the call to `statistics(&self)`?

**Comment by Jimexist** [comment]
> I think don't think adding an RwLock sounds like a good idea. Perhaps we could compute the statistics when the `ValuesExec` is created (rather than computing on the call to `statistics(&self)`?

agreed, no need for a lock. we can just put evaluation in the creation part of `ValuesExec`. i mean it probably wouldn't make any difference

**Comment by ygf11** [comment]
@alamb IMOP, earlier computing statistics conflicts with the purpose of this task, because this will iterate all row and column, which also slow the creation of ValueExec. Maybe stay the same is ok? 

> I think don't think adding an RwLock sounds like a good idea. Perhaps we could compute the statistics when the ValuesExec is created (rather than computing on the call to statistics(&self)?

**Comment by haochunchang** [comment]
Following up here since this has sat open since 2021 with an unresolved design question.

ValuesExec no longer exists in the codebase. It was deprecated in favor of MemoryExec (#14032), folded into the unified DataSourceExec/MemorySourceConfig abstraction (#14224), and the standalone values.rs file was deleted outright as part of the v45 deprecation cleanup (#17075).

VALUES-list handling now lives in [MemorySourceConfig::try_new_as_values](https://github.com/apache/datafusion/blob/main/datafusion/datasource/src/memory.rs#L377). 
It still evaluates the literal expressions into a RecordBatch eagerly, at construction time.
`execute()/open()` just clones the pre-built batches, and statistics()/partition_statistics() reads from that same eager data.

I'd propose to close this issue. Happy to learn for what I have missed.

**Comment by alamb** [comment]
Thgank you @haochunchang  -- sounds good to me
